### PR TITLE
add `prep_nisar` to load NISAR GUNW products

### PR DESCRIPTION
**Pull request by mirzaees**
**Description of proposed changes**

Add prep_nisar.py script to read GUNW products. The inputs are a list of nisar*.h5 data cubes and the DEM.

Future to do list:
- The water mask in the test products is all zero. To avoid masking the data, it was manually set to one. We need to remove manual part after correctly adding water mask to input data.
- Tropospheric correction of Mintpy using pyaps does not perform well because of UTM coordinate system. We might read the correction layers from input data instead in the future. But for now, the height_correlation option works.
- The x/y coordinates of data cube in radarGrid require to add a margin to cover the x/y coordinates of GUNW. For now the common bounding box to read all data is the area covered by both radarGrid and GUNW grid, meaning we cut the outer edges if needed. 
- The auto path for the input products is set to "../interferograms/*.h5". It might change based on where the GUNW products are located after processing with isce3

### CODE REVIEW

**Review by scottstanie — COMMENTED** [review]
I'm curious what @yunjunz will think about the radar grid/geo-grid problem... the `incidence`/`bperp` being on a different grid is why Sara had to add all the extra checks/interpolation/creation of a larger bounding box that each product is inside. 
- It's probably necessary that the `geometry.h5` file is on the same geocoded grid as the `ifgramStack.h5`, right?  Maybe we could just warp the geometry ones to be on the same as the GUNW grid using `gdalwarp`?

File: src/mintpy/prep_nisar.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    for key, value in DATASETS.items():
+        if value:
+            value = value.replace("POL", polarization)
+            DATASETS[key] = value
+
+    with h5py.File(meta_file, 'r') as ds:
+        pixel_height = ds[DATASETS['ySpacing']][()]
+        pixel_width = ds[DATASETS['xSpacing']][()]
+        x_origin = min(ds[DATASETS['xcoord']][()])
+        y_origin = min(ds[DATASETS['ycoord']][()])
```
[the remark this answers is not in the document]
Comment (mirzaees) [reply src/mintpy/prep_nisar.py]: Thank you @scottstanie , I believe we only need to correct this line

File: src/mintpy/cli/prep_nisar.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+# Author: Sara Mirzaee, Jul 2023                           #
+############################################################
+
+import sys
+
+from mintpy.utils.arg_utils import create_argument_parser
+
+############################################################
+EXAMPLE = """example:
+  python3 ./prep_nisar.py -i 'interferograms/stitched/*.h5' -d 'dem.tiff'
```
Comment (yunjunz) [inline src/mintpy/cli/prep_nisar.py:14]:
```suggestion
  prep_nisar.py -i 'interferograms/stitched/*.h5' -d dem.tiff
```

File: src/mintpy/prep_nisar.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+
+    t_mid = start_time + (end_time - start_time) / 2.0
+    meta["CENTER_LINE_UTC"] = (
+            t_mid - datetime.datetime(t_mid.year, t_mid.month, t_mid.day)
+    ).total_seconds()
+   
+    meta["X_FIRST"] = x_origin - pixel_width//2
+    meta["Y_FIRST"] = y_origin - pixel_height//2
+    meta["X_STEP"] = pixel_width
+    meta["Y_STEP"] = pixel_height
```
[the remark this answers is not in the document]
Comment (yunjunz) [reply src/mintpy/prep_nisar.py:138]: `Y_STEP` should be negative, `X_STEP` should be positive, and `Y/X_FIRST` represents the upper left corner of the upper left pixel. Please ensure this.

**Review by yunjunz — COMMENTED** [review]
Thank you @mirzaees for the exciting PR! Sorry for my very late review, I got distracted by the EarthScope course and proposal writings. 

I don't really have a good grasp on the radar/geo-grid of the geometry files that @scottstanie was mentioning. Is there an open interferogram stack in the NISAR format, or a detailed description of the InSAR file format, that we could test? Otherwise, the PR is fine with me to merge since it does not break any other codes.

File: src/mintpy/prep_nisar.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mirzaees) [reply src/mintpy/prep_nisar.py:138]: great, it is as you mentioned @yunjunz

**Comment by mirzaees** [comment]
Thank you @yunjunz for reviewing 
I do not have an open interferogram stack but I have tested this, and confirm it is working. So let's merge it if it is fine with you.

**Review by yunjunz — APPROVED** [review]

**Pull request merged**